### PR TITLE
W-11177824: flow ref validations fails for xml sdk components that declare inner flows when using lazy init

### PR DESCRIPTION
**Pull request by SebaElizalde**

### CODE REVIEW

File: modules/extensions-xml-support/src/main/java/org/mule/runtime/extension/internal/ast/MacroExpansionModulesModel.java [hunk 1]
```diff
@@ -86,6 +93,33 @@ public ArtifactAst expand() {
...
 
     if (hasMacroExpansionExtension) {
+      // The macro expansion of xml sdk components causes inner flow components to be generated as a child of another component;
+      // so when generating the minimal app during lazy init, those flow components are not registered as top level components
+      // causing certain validations to fail. This code looks for those flow components and adds them as top level.
+      Stream<ComponentAst> flowComponents =
+          applicationModel.recursiveStream()
+              .filter(comp -> comp.getComponentType() != null && comp.getComponentType().equals(FLOW));
+      List<ComponentAst> componentsToAdd =
+          flowComponents.filter(comp -> !applicationModel.topLevelComponents().contains(comp)).collect(Collectors.toList());
```
Comment (elrodro83) [inline modules/extensions-xml-support/src/main/java/org/mule/runtime/extension/internal/ast/MacroExpansionModulesModel.java:103]: inline `flowComponents`

File: modules/extensions-xml-support/src/main/java/org/mule/runtime/extension/internal/ast/MacroExpansionModulesModel.java [hunk 2]
```diff
@@ -86,6 +93,33 @@ public ArtifactAst expand() {
     }
 
     if (hasMacroExpansionExtension) {
+      // The macro expansion of xml sdk components causes inner flow components to be generated as a child of another component;
+      // so when generating the minimal app during lazy init, those flow components are not registered as top level components
+      // causing certain validations to fail. This code looks for those flow components and adds them as top level.
+      Stream<ComponentAst> flowComponents =
+          applicationModel.recursiveStream()
+              .filter(comp -> comp.getComponentType() != null && comp.getComponentType().equals(FLOW));
```
Comment (elrodro83) [inline modules/extensions-xml-support/src/main/java/org/mule/runtime/extension/internal/ast/MacroExpansionModulesModel.java:101]: be careful to only move flows from the extension that is being macroexpanded

File: modules/extensions-xml-support/src/main/java/org/mule/runtime/extension/internal/ast/MacroExpansionModulesModel.java (hunk 1 above)
Comment (elrodro83) [inline modules/extensions-xml-support/src/main/java/org/mule/runtime/extension/internal/ast/MacroExpansionModulesModel.java:103]: `!applicationModel.topLevelComponents().contains(comp)` shouldn't be needed if you are only navigating the extension being macroexpanded

Comment (elrodro83) [inline modules/extensions-xml-support/src/main/java/org/mule/runtime/extension/internal/ast/MacroExpansionModulesModel.java:103]: static import: `Collectors.toList()`

File: modules/extensions-xml-support/src/main/java/org/mule/runtime/extension/internal/ast/MacroExpansionModulesModel.java [hunk 3]
```diff
@@ -86,6 +93,33 @@ public ArtifactAst expand() {
...
+      List<ComponentAst> componentsToAdd =
+          flowComponents.filter(comp -> !applicationModel.topLevelComponents().contains(comp)).collect(Collectors.toList());
+      final UnaryOperator<ComponentAst> function = comp -> {
+        List<ComponentAst> childrenToKeep = comp.directChildrenStream()
+            .filter(child -> !(child.getComponentType() != null && child.getComponentType().equals(FLOW)))
+            .collect(Collectors.toList());
```
Comment (elrodro83) [inline modules/extensions-xml-support/src/main/java/org/mule/runtime/extension/internal/ast/MacroExpansionModulesModel.java:107]: static import

File: modules/extensions-xml-support/src/main/java/org/mule/runtime/extension/internal/ast/MacroExpansionModulesModel.java [hunk 4]
```diff
@@ -86,6 +93,33 @@ public ArtifactAst expand() {
...
+      // so when generating the minimal app during lazy init, those flow components are not registered as top level components
+      // causing certain validations to fail. This code looks for those flow components and adds them as top level.
+      Stream<ComponentAst> flowComponents =
+          applicationModel.recursiveStream()
+              .filter(comp -> comp.getComponentType() != null && comp.getComponentType().equals(FLOW));
+      List<ComponentAst> componentsToAdd =
+          flowComponents.filter(comp -> !applicationModel.topLevelComponents().contains(comp)).collect(Collectors.toList());
+      final UnaryOperator<ComponentAst> function = comp -> {
+        List<ComponentAst> childrenToKeep = comp.directChildrenStream()
+            .filter(child -> !(child.getComponentType() != null && child.getComponentType().equals(FLOW)))
```
Comment (elrodro83) [inline modules/extensions-xml-support/src/main/java/org/mule/runtime/extension/internal/ast/MacroExpansionModulesModel.java:106]: `!(child.getComponentType() != null && child.getComponentType().equals(FLOW))` -> `!FLOW.equals(child.getComponentType())`

File: modules/extensions-xml-support/src/main/java/org/mule/runtime/extension/internal/ast/MacroExpansionModulesModel.java [hunk 5]
```diff
@@ -86,6 +93,33 @@ public ArtifactAst expand() {
...
+            return childrenToKeep;
+          }
+
+          @Override
+          public Stream<ComponentAst> directChildrenStream() {
+            return childrenToKeep.stream();
+          }
+        };
+      };
+      applicationModel = copyRecursively(applicationModel, function, () -> componentsToAdd, c -> false);
```
Comment (elrodro83) [inline modules/extensions-xml-support/src/main/java/org/mule/runtime/extension/internal/ast/MacroExpansionModulesModel.java:121]: inline `function`

**Review by elrodro83 — CHANGES_REQUESTED** [review]

File: modules/extensions-xml-support/src/main/java/org/mule/runtime/extension/internal/ast/MacroExpansionModulesModel.java (hunk 2 above)
Comment (SebaElizalde) [reply modules/extensions-xml-support/src/main/java/org/mule/runtime/extension/internal/ast/MacroExpansionModulesModel.java:101]: all extensions have been macroexpanded and added to the application model by this point

File: modules/extensions-xml-support/src/main/java/org/mule/runtime/extension/internal/ast/MacroExpansionModulesModel.java (hunk 1 above)
Comment (SebaElizalde) [reply modules/extensions-xml-support/src/main/java/org/mule/runtime/extension/internal/ast/MacroExpansionModulesModel.java:103]: Even if dealing with only 1 extension, there is the case that the extension has 2 or more operations pointing to the same flow, so while travelling the ast you can find the flow twice, but you only want to add it once to top level. So this check should remain imo.

**Review by elrodro83 — APPROVED** [review]

**Comment by SebaElizalde** [comment]
--validate

**Pull request merged**